### PR TITLE
Add Advanced Security Alert tools

### PR DESCRIPTION
**Pull request by nuthanmunaiah**
We are introducing two new tools---`advsec_get_alerts` and 
`advsec_get_alert_details`---in this change to allow customers to
work with [GHAzDO (GitHub Advanced Security for Azure
DevOps)](https://aka.ms/advanced-security) alerts directly in code 
editor.

## GitHub issue number

N/A

## **Associated Risks**

None

## ✅ **PR Checklist**

- [x] **I have read the [contribution guidelines](https://github.com/microsoft/azure-devops-mcp/blob/main/CONTRIBUTING.md)**
- [x] **I have read the [code of conduct guidelines](https://github.com/microsoft/azure-devops-mcp/blob/main/CODE_OF_CONDUCT.md)**
- [x] Title of the pull request is clear and informative.
- [x] 👌 Code hygiene
- [ ] 🔭 ~~Telemetry added, updated, or N/A~~ N/A
- [x] 📄 Documentation added, updated, or N/A
- [x] 🛡️ Automated tests added, or N/A

## 🧪 **How did you test it?**

  * Automated unit tests run using `npm run test`
  * Manual testing with VS Code Agent Mode in Copilot Chat.

### CODE REVIEW

**Comment by danhellem** [comment]
@nuthanmunaiah we don't yet have an owner for the GHAzDO area and may not want to merge anything around that space until we do. Stand by.

**Comment by danhellem** [comment]
@nuthanmunaiah also, in the future. Please follow the contribution guidelines before submitting a PR.

https://github.com/microsoft/azure-devops-mcp/blob/main/CONTRIBUTING.md#-writing-code

File: src/tools/advsec.ts [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      onlyDefaultBranch: z.boolean().optional().default(true).describe("If true, only return alerts found on the default branch. Defaults to true."),
+      confidenceLevels: z
+        .array(z.enum(["high", "other"]))
+        .optional()
+        .describe("Filter alerts by confidence levels. Only applicable for secret alerts."),
+      validity: z
+        .array(z.enum(["active", "inactive", "none", "unknown"]))
+        .optional()
+        .describe("Filter alerts by validity status. Only applicable for secret alerts."),
+      top: z.number().optional().default(100).describe("Maximum number of alerts to return. Defaults to 100."),
```
Comment (laurajjiang) [inline src/tools/advsec.ts:49]: Can you confirm that 100 alerts with full metadata doesn't overwrite the Copilot context window for you? In my tool implementation, I had to cut down to top ~50 as well as stripping some extraneous metadata.

File: src/tools/advsec.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+      orderBy: z.enum(["id", "firstSeen", "lastSeen", "fixedOn", "severity"]).optional().default("id").describe("Order results by specified field. Defaults to 'id'."),
```
Comment (laurajjiang) [inline src/tools/advsec.ts]: Can we default order by severity instead? I don't think most customers would sort by id, nor do we by default for any of our UI scenarios.

File: src/tools/advsec.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+      severities: z
+        .array(z.enum(["critical", "high", "medium", "low", "note", "error", "warning", "undefined"]))
+        .optional()
+        .describe("Filter alerts by severity level. If not specified, returns alerts at any severity."),
+      ruleId: z.string().optional().describe("Filter alerts by rule ID."),
+      ruleName: z.string().optional().describe("Filter alerts by rule name."),
+      toolName: z.string().optional().describe("Filter alerts by tool name."),
+      ref: z.string().optional().describe("Filter alerts by git reference (branch). If not provided and onlyDefaultBranch is true, only includes alerts from default branch."),
+      onlyDefaultBranch: z.boolean().optional().default(true).describe("If true, only return alerts found on the default branch. Defaults to true."),
+      confidenceLevels: z
```
Comment (laurajjiang) [inline src/tools/advsec.ts:40]: Can we also add that by default, we fetch both? This has been a point of annoyance from some customers that we only return high confidence alerts.

File: src/tools/advsec.ts (hunk 1 above)
Comment (nuthanmunaiah) [reply src/tools/advsec.ts:49]: How does one know if the Copilot context window is full? I used the server to get all 549 active dependency scanning alerts for one of our test repositories, and Copilot was able to summarize the results just fine.

Comment (laurajjiang) [reply src/tools/advsec.ts:49]: If the tool call is successfully processed and summarized after the call, then we should be good. In my testing, the tool call would continuously loop, which I learned meant was because it surpassed the context window. Sounds like this was patched/changed in some way.

File: src/tools/advsec.ts (hunk 3 above)
Comment (nuthanmunaiah) [reply src/tools/advsec.ts:40]: Done. Unfortunately, however, this change will not work as expected because I discovered a bug in `azure-devops-node-api` which causes the confidence levels of `["high", "other"]` to be translated to `criteria.confidenceLevels.0=0&criteria.confidenceLevels.1=1` instead of `criteria.confidenceLevels=0&criteria.confidenceLevels=1`. Since the service does not recognize `criteria.confidenceLevels.n` as a valid query parameter, it defaults to returning high confidence secret alerts only.

File: src/tools/advsec.ts (hunk 2 above)
Comment (nuthanmunaiah) [reply src/tools/advsec.ts]: Done.

File: src/tools/advsec.ts (hunk 3 above)
Comment (laurajjiang) [reply src/tools/advsec.ts:40]: Should we remove this criteria and add a comment in code that we only currently return high confidence via this tool? I don't want the wrong impression that this would work, then get feedback / issues from customers.

Comment (nuthanmunaiah) [reply src/tools/advsec.ts:40]: I noticed that the bug applies to every filter criterion that allows multiple values. For instance, if I want to reason on fixed alerts, the query parameter for state is getting set to `criteria.states.0=4` (`4` here is fixed) instead of `criteria.states=4`. Let me reach out to `azure-devops-node-api` Team to figure out if there is anything we can do to mitigate this.

**Comment by nuthanmunaiah** [comment]
@polatengin I added the two functions you recommended to `src/utils.ts`. I also added unit tests to `test/src/utils.test.ts` to test these two new functions..

Note: While I was working on the change, it appears two new similar functions (`getEnumKeys` and `safeEnumConvert`) were added to `src/utils.ts`. `safeEnumConvert` function, however, does not transform an enumeration to a `Record<string, T[keyof T]>` before attempting to map a string to an enumeration value. A side effect of this implementation choice is that the `key` argument to `safeEnumConvert` must be case sensitive.

**Example**

```TypeScript
enum AlertType { Unknown = 0, Dependency = 1, Secret = 2, Code = 3, License = 4 }

console.log(safeEnumConvert(AlertType, "Code")); // Returns `3`
console.log(safeEnumConvert(AlertType, "code")); // Returns `undefined`
```

File: src/tools/advsec.ts (hunk 3 above)
Comment (nuthanmunaiah) [reply src/tools/advsec.ts:40]: Opened microsoft/azure-devops-node-api#644 to fix the bug.

File: src/tools/advsec.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+function configureAdvSecTools(server: McpServer, tokenProvider: () => Promise<AccessToken>, connectionProvider: () => Promise<WebApi>) {
+  server.tool(
+    ADVSEC_TOOLS.get_alerts,
+    "Retrieve Advanced Security alerts for a repository.",
+    {
+      project: z.string().describe("The name or ID of the Azure DevOps project."),
+      repository: z.string().describe("The name or ID of the repository to get alerts for."),
+      alertType: z.enum(["code", "dependency", "secret", "unknown"]).optional().describe("Filter alerts by type. If not specified, returns all alert types."),
+      states: z
+        .array(z.enum(["active", "dismissed", "fixed", "autoDismissed", "unknown"]))
```
Comment (Novaes) [inline src/tools/advsec.ts]: Can we use the typescript sdk enums here?  We have some examples on repos.

File: src/tools/advsec.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+};
+
+function configureAdvSecTools(server: McpServer, tokenProvider: () => Promise<AccessToken>, connectionProvider: () => Promise<WebApi>) {
+  server.tool(
+    ADVSEC_TOOLS.get_alerts,
+    "Retrieve Advanced Security alerts for a repository.",
+    {
+      project: z.string().describe("The name or ID of the Azure DevOps project."),
+      repository: z.string().describe("The name or ID of the repository to get alerts for."),
+      alertType: z.enum(["code", "dependency", "secret", "unknown"]).optional().describe("Filter alerts by type. If not specified, returns all alert types."),
```
Comment (Novaes) [inline src/tools/advsec.ts]: same here for using typescript sdk enums

File: src/tools/advsec.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      states: z
+        .array(z.enum(["active", "dismissed", "fixed", "autoDismissed", "unknown"]))
+        .optional()
+        .describe("Filter alerts by state. If not specified, returns alerts in any state."),
+      severities: z
+        .array(z.enum(["critical", "high", "medium", "low", "note", "error", "warning", "undefined"]))
```
Comment (Novaes) [inline src/tools/advsec.ts]: same for those

File: src/tools/advsec.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      top: z.number().optional().default(100).describe("Maximum number of alerts to return. Defaults to 100."),
+      orderBy: z.enum(["id", "firstSeen", "lastSeen", "fixedOn", "severity"]).optional().default("severity").describe("Order results by specified field. Defaults to 'severity'."),
+      continuationToken: z.string().optional().describe("Continuation token for pagination."),
+    },
+    async ({ project, repository, alertType, states, severities, ruleId, ruleName, toolName, ref, onlyDefaultBranch, confidenceLevels, validity, top, orderBy, continuationToken }) => {
+      try {
+        const connection = await connectionProvider();
+        const alertApi = await connection.getAlertApi();
+
+        const criteria = {
```
Comment (Novaes) [inline src/tools/advsec.ts:59]: This is: 
* Mapping string inputs to enum values using mapStringToEnum(...).
* Filtering out undefined values to ensure only valid enum entries are included.
* Conditionally including fields only if their corresponding input exists.

this is quite complex, can we move this to a separate class /utils or even at least just a function, to make it easy to maintain this piece of logic.

File: src/tools/advsec.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      confidenceLevels: z
+        .array(z.enum(["high", "other"]))
+        .optional()
+        .default(["high", "other"])
+        .describe("Filter alerts by confidence levels. Only applicable for secret alerts. Defaults to both 'high' and 'other'."),
+      validity: z
+        .array(z.enum(["active", "inactive", "none", "unknown"]))
+        .optional()
+        .describe("Filter alerts by validity status. Only applicable for secret alerts."),
+      top: z.number().optional().default(100).describe("Maximum number of alerts to return. Defaults to 100."),
```
Comment (Novaes) [inline src/tools/advsec.ts:49]: Can you confirm pagination working properly? If yes, I'd even start with lower number, 50 for example.

File: src/utils.ts [hunk 9]
```diff
@@ -7,6 +7,21 @@ export const apiVersion = "7.2-preview.1";
 export const batchApiVersion = "5.0";
 export const markdownCommentsApiVersion = "7.2-preview.4";
 
+export function createEnumMapping<T extends Record<string, string | number>>(enumObject: T, keyTransform?: (key: string) => string): Record<string, T[keyof T]> {
```
Comment (Novaes) [inline src/utils.ts]: this seems quite error prone and complex. Can we simplify it? An example: https://stackoverflow.com/questions/41308123/map-typescript-enum

File: test/src/tools/advsec.test.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+  });
+
+  describe("tool registration", () => {
+    it("registers Advanced Security tools on the server", () => {
+      configureAdvSecTools(server, tokenProvider, connectionProvider);
+      expect(server.tool as jest.Mock).toHaveBeenCalled();
+    });
+  });
+
+  describe("advsec_get_alerts tool", () => {
```
Comment (Novaes) [inline test/src/tools/advsec.test.ts:47]: missing testing with multiple alerts

File: test/src/utils.test.ts [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+      });
+
+      it("should return undefined for invalid strings without default", () => {
+        expect(mapStringToEnum("invalid", alertTypeMapping)).toBeUndefined();
+        expect(mapStringToEnum("nonexistent", alertTypeMapping)).toBeUndefined();
+      });
+    });
+
+    describe("edge cases", () => {
+      it("should handle undefined input", () => {
```
Comment (Novaes) [inline test/src/utils.test.ts:115]: Consider clarifying what handle means here: 
`when passing undefined input should return unkown`

File: test/src/utils.test.ts [hunk 12]
```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
```
Comment (Novaes) [inline test/src/utils.test.ts:3]: consider adding some video or adding some use cases validated for the: 
>Manual testing with VS Code Agent Mode in Copilot Chat.

while we work on integration and e2e tests. Thanks in advance for it.

**Review by Novaes — APPROVED** [review]
LGTM with comments 👍

File: src/tools/advsec.ts (hunk 8 above)
Comment (nuthanmunaiah) [reply src/tools/advsec.ts:49]: Unfortunately, there appears to be an issue in `azure-devops-node-api` that is preventing pagination. See microsoft/azure-devops-node-api#646 for the details. We will leave this at 100 for now, but follow up with a change to lower it after the issue is fixed.

File: test/src/utils.test.ts (hunk 12 above)
Comment (nuthanmunaiah) [reply test/src/utils.test.ts:3]: I am working on a demo of the end-to-end experience of using the two new tools. I will share the recording here after I have had someone review the video to strip any Microsoft-internal content from the video.

File: test/src/tools/advsec.test.ts (hunk 10 above)
Comment (nuthanmunaiah) [reply test/src/tools/advsec.test.ts:47]: Added test for multiple alerts.

File: test/src/utils.test.ts (hunk 11 above)
Comment (nuthanmunaiah) [reply test/src/utils.test.ts:115]: We test the following two cases when `value` to map to an enumeration is `undefined`:

  1. When there is no `defaultValue` parameter specified, `undefined` should be returned.
  2. When a `defaultValue` of `AlertType.Unknown` is specified, `AlertType.Unknown` should be returned.

Does this help clarify the intent here?

File: src/utils.ts (hunk 9 above)
Comment (nuthanmunaiah) [reply src/utils.ts]: I simplified it by removing the `keyTransform` argument. The implementation of the `createEnumMapping` is largely inspired by [earlier comment](https://github.com/microsoft/azure-devops-mcp/pull/270#pullrequestreview-3040271177) from @polatengin.
